reactive/ntpmon.py

```diff
@@ -56,53 +74,79 @@ def install_ntpmon():
-        hookenv.log('installing ntpmon systemd configuration')
-        host.rsync('src/' + service_name + '.systemd', systemd_config)
-        subprocess.call(['systemd', 'daemon-reload'])
-    else:
-        hookenv.log('installing ntpmon upstart configuration')
-        host.rsync('src/' + service_name + '.upstart', upstart_config)
+    install_dir = get_option('install-dir')
+    service_name = get_option('service-name')
+    using_systemd = host.init_is_systemd()
+    if install_dir:
+        log('installing ntpmon')
+        host.mkdir(os.path.dirname(install_dir))
+        host.rsync('src/', '{}/'.format(install_dir))
+
+        if service_name:
+            if using_systemd:
+                systemd_config = '/etc/systemd/system/' + service_name + '.service'
+                log('installing systemd service: {}'.format(service_name))
+                with open(systemd_config, 'w') as conffile:
+                    conffile.write(templating.render('src/' + service_name + '.systemd', ntpmon_options))
+                subprocess.call(['systemd', 'daemon-reload'])
+            else:
+                upstart_config = '/etc/init/' + service_name + '.conf'
+                log('installing upstart service: {}'.format(service_name))
+                with open(upstart_config, 'w') as conffile:
+                    conffile.write(templating.render('src/' + service_name + '.upstart', ntpmon_options))
+
     set_state('ntpmon.installed')
     remove_state('ntpmon.configured')
 
 
+# TODO: implement removal
+
+
 @when('ntpmon.installed')
 @when_not('ntpmon.configured')
 def configure_ntpmon():
     """
     Reconfigure ntpmon - does nothing at present
     """
-    hookenv.log('configuring ntpmon')
+    log('configuring ntpmon')
     set_state('ntpmon.configured')
     remove_state('ntpmon.started')
 
 
-@when('ntpmon.configured', 'telegraf.configured')
+@when('ntpmon.configured')
 @when_not('ntpmon.started')
 def start_ntpmon():
     """
     Start the ntpmon daemon process.
-    If ntp is not installed, do nothing.
+    If no NTP server is installed, do nothing.
     """
-    if os.path.exists(ntp_conf):
-        hookenv.log(ntp_conf + ' present; enabling and starting ntpmon')
-        host.service_resume(service_name)
-    else:
-        hookenv.log(ntp_conf + ' not present; disabling ntpmon')
-        host.service_pause(service_name)
+    service_name = get_option('service-name')
+    started = False
+    if service_name:
+        for f in (chrony_conf, ntp_conf):
+            if os.path.exists(f):
+                log('{} present; enabling and starting ntpmon'.format(f))
+                host.service_resume(service_name)
+                started = True
+                break
+        if not started:
+            log('No supported NTP service present; disabling ntpmon')
+            host.service_pause(service_name)
     set_state('ntpmon.started')
 
 
+# TODO: implement stop
+
+
+@when_file_changed([chrony_conf], hash_type='sha256')
```

Comment:
I could have sworn when_file_changed got deprecated, because there are ways for the handler to get skipped. Personally, I'd just clear the ntpmon.started flag after writing the config file (using charms.reactive.helpers.any_file_changed if you want to catch noops).

Reply:
This layer doesn't touch the config file at all, and neither does the telegraf charm which includes it; only the ntp charm does.  The current process is in place to detect the situation where the NTP implementation is changed, either manually by the user, or automatically by the ntp charm during a series upgrade.  I'm open to better ways of doing this, but this seemed to be most sensible way to handle it, and when I tested it, it did work.  Also, https://charmsreactive.readthedocs.io/en/latest/charms.reactive.decorators.html#charms.reactive.decorators.when_file_changed doesn't list it as deprecated.